roxie/ccd/ccdcontext.cpp

```diff
@@ -3710,9 +3784,33 @@ class CRoxieServerContext : public CRoxieContextBase, implements IRoxieServerCon
     virtual IConstWorkUnit *queryWorkUnit() const { return workUnit; }
     virtual bool outputResultsToSocket() const { return client != NULL; }
 
-    virtual void selectCluster(const char * cluster) { throwUnexpected(); }
-    virtual void restoreCluster() { throwUnexpected(); }
-
+    virtual void selectCluster(const char * newCluster)
+    {
+        if (workUnit)
+        {
+            const char *oldCluster = workUnit->queryClusterName();
+            SCMStringBuffer bStr;
+            ClusterType targetClusterType = getClusterType(workUnit->getDebugValue("targetClusterType", bStr).str(), RoxieCluster);
+            if (targetClusterType==RoxieCluster)
+            {
+                if (!streq(oldCluster, newCluster))
+                    throw MakeStringException(-1, "Error - cannot switch cluster if not targetting thor jobs");
+            }
+            clusterNames.append(oldCluster);
```

Comment:
This seems wrong - all the code above is assuming that clusterNames.tos() is the current cluster - but this is pushing the old cluster (which is already on the list for the 1st entry).